shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -308,31 +308,36 @@ - (void)removeSplashScreenView:(dispatch_block_t _Nullable)onComplete {
 }
 
 - (void)installFirstFrameCallback {
-  auto weak_platform_view = [_engine.get() platformView];
+  fml::WeakPtr<flutter::PlatformViewIOS> weak_platform_view = [_engine.get() platformView];
   if (!weak_platform_view) {
     return;
   }
-  __unsafe_unretained auto weak_flutter_view_controller = self;
+
   // This is on the platform thread.
-  weak_platform_view->SetNextFrameCallback([weak_platform_view, weak_flutter_view_controller,
-                                            task_runner = [_engine.get() platformTaskRunner]]() {
-    // This is on the GPU thread.
-    task_runner->PostTask([weak_platform_view, weak_flutter_view_controller]() {
-      // We check if the weak platform view is alive. If it is alive, then the view controller
-      // also has to be alive since the view controller owns the platform view via the shell
-      // association. Thus, we are not convinced that the unsafe unretained weak object is in
-      // fact alive.
-      if (weak_platform_view) {
-        if (weak_flutter_view_controller->_splashScreenView) {
-          [weak_flutter_view_controller removeSplashScreenView:^{
-            [weak_flutter_view_controller callViewRenderedCallback];
-          }];
-        } else {
-          [weak_flutter_view_controller callViewRenderedCallback];
-        }
-      }
-    });
-  });
+  weak_platform_view->SetNextFrameCallback(
+      [weak_platform_view, platform_task_runner = [_engine.get() platformTaskRunner],
+       gpu_task_runner = [_engine.get() gpuTaskRunner]]() {
+        FML_DCHECK(gpu_task_runner->RunsTasksOnCurrentThread());
+        platform_task_runner->PostTask([weak_platform_view]() {
+          // The weak platform view will have a weak pointer to self, if it is
+          // still alive. We can check if that weak pointer is still valid to
+          // know if we're still alive.
+          if (weak_platform_view) {
+            fml::WeakPtr<FlutterViewController> weak_self =
+                weak_platform_view->GetOwnerViewController();
+            if (!weak_self) {
+              return;
+            }
+            if (weak_self.get()->_splashScreenView) {
```

Comment:
weak_self and weak_platform_view could be released after their nil checks.  Retain them before you use them, then release when you're done with them.

Reply:
They can't be released here because we're on the queue that is responsible for them and we're not yielding.  

I'm not sure how I could retain them - the `get()` just gives me a raw pointer that I can't retain.

Reply:
I don't know how these fml::WeakPtr work, I'm going to defer to someone who knows this code better than I do.

Reply:
> a raw pointer that I can't retain

I don't follow; if you have a pointer to an ObjC object, then you can retain it. It's best practice to retain objects across making multiple calls to them (as jmagman is suggesting) in case one of the calls along the way would otherwise cause it to be deallocated as a side effect.

Reply:
Ahh ok, I can retain the view controller but not the platform view. Is that better?